posts/2019-07-15-Rust-1.37.0.md

```diff
@@ -0,0 +1,176 @@
+
+### Referring to `enum` variants through `type` aliases
+
+With Rust 1.37.0, you can now refer to `enum` variants through type aliases. For example:
+
+```rust
+type ByteOption = Option<u8>;
+
+fn increment_or_zero(x: ByteOption) -> u8 {
+    match x {
+        ByteOption::Some(y) => y + 1,
+        ByteOption::None => 0,
+    }
+}
+```
+
+In implementations, `Self` acts like a type alias. So in Rust 1.37.0, you can also refer to `enum` variants with `Self::Variant`:
+
+```rust
+impl Coin {
+    fn value_in_cents(&self) -> u8 {
+        match self {
+            Self::Penny => 1,
+            Self::Nickel => 5,
+            Self::Dime => 10,
+            Self::Quarter => 25,
+        }
+    }
+}
+```
+
+[type_rel_report]: https://github.com/rust-lang/rust/pull/61682/#issuecomment-502472847
+
+To be more exact, Rust now allows you to refer to `enum` variants through *"type-relative resolution"*, `<MyType<..>>::Variant`. More details are available in [the stabilization report][type_rel_report].
+
+### Built-in Cargo support for vendored dependencies
+
+[vendor-crate]: https://crates.io/crates/cargo-vendor
+
+After being available [as a separate crate][vendor-crate] for years, the `cargo vendor` command is now integrated directly into Cargo. The command fetches all your project's dependencies unpacking them into the `vendor/` directory, and shows the configuration snippet required to use the vendored code during builds.
+
+There are multiple cases where `cargo vendor` is already used in production: the Rust compiler `rustc` uses it to ship all its dependencies in release tarballs, and projects with monorepos use it to commit the dependencies' code in source control.
+
+### Using unnamed `const` items for macros
+
+[unnamed_const_pr]: https://github.com/rust-lang/rust/pull/61347/
+
+You can now create [unnamed `const` items][unnamed_const_pr]. Instead of giving your constant an explicit name, simply name it `_` instead. For example, in the `rustc` compiler we find:
+
+```rust
+/// Type size assertion where the first parameter
+/// is a type and the second is the expected size.
+#[macro_export]
+macro_rules! static_assert_size {
+    ($ty:ty, $size:expr) => {
+        const _: [(); $size] = [(); ::std::mem::size_of::<$ty>()];
+        //    ^ Note the underscore here.
+    }
+}
+
+static_assert_size!(Option<Box<String>>, 8); // 1.
+static_assert_size!(usize, 8); // 2.
+```
+
+Notice the second `static_assert_size!(..)`: thanks to the use of unnamed constants, you can define new items without naming conflicts. Previously you would have needed to write `static_assert_size!(MY_DUMMY_IDENTIFIER, usize, 8);`. Instead, with Rust 1.37.0, it now becomes easier to create ergonomic and reusable declarative and procedural macros for static analysis purposes.
```

Comment:
Could anyone comment on why this is not solved by hygiene? E.g., if you wrote `const MY_DUMMY_IDENTIFIER: ...` in the body of the macro, I would expect hygiene to prevent naming conflicts

Reply:
What hygiene? =P

```rust
macro_rules! foo {
    () => {
        const MY_DUMMY_IDENTIFIER: () = ();
    }
}

foo!();
//foo!(); -- comment out ==> error.

fn main() {
    let _ = MY_DUMMY_IDENTIFIER; // OK
}
```

Macros 2.0 are still not a thing. ;)

Reply:
Ah, right, `const`s are items not variables so they don't get hygiene :-(